### PR TITLE
IDE: Operation-specific metrics and small improvements

### PR DESCRIPTION
**Pull request by jvilk-stripe**
IDE: Operation-specific metrics and small improvements

Specific changes:

* `LSPMessage`s now have a second optional timer that tracks the latency of specific methods.
* `LSPMessage` now contains message (and timer) cancelation logic.
* `enableMultithreading` is now a property of `LSPTask` rather than a parameter to `TypecheckerCoordinator::syncRun`.
* Renamed the confusing `TypecheckerTask::hasDedicatedThread` to `TypecheckerTask::collectCounters`.

### Motivation

We have no way to check the latency of specific operations in SignalFX, as all operations are grouped together 

### Test plan

See included automated tests.

### CODE REVIEW

File: main/lsp/LSPMessage.cc [hunk 1]
```diff
@@ -113,6 +113,23 @@ LSPMessage::LSPMessage(rapidjson::Document &d) : LSPMessage::LSPMessage(fromJSON
...
 LSPMessage::LSPMessage(const std::string &json) : LSPMessage::LSPMessage(fromJSON(json)) {}
 
+bool LSPMessage::isCanceled() const {
+    return canceled;
+}
+
+void LSPMessage::cancel() {
+    if (!canceled) {
+        canceled = true;
+        // Protect against nullptrs.
```
Comment (mkillianey-stripe) [inline main/lsp/LSPMessage.cc]: Personal preference nit: although I'm generally in favor of more comments, this specific one didn't give me any extra information that wasn't clear to me from the code.  (I see that this code moved, and the original had this comment, so just giving you the liberty to remove if you found it equally tame.)

Comment (jvilk-stripe) [reply main/lsp/LSPMessage.cc]: Ah, I can add context: A Timer may not be specified (it's optional).

**Comment by jvilk-stripe** [comment]
why is this requesting reviews? sorry about that

**Comment by jez** [comment]
I think dmitry changed some things.

**Pull request merged**